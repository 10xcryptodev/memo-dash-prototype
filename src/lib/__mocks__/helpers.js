import testUsers from '../../test-utils/test-users'

export const filterUser = (username, users) => testUsers.find(testUser => testUser.username === username)
