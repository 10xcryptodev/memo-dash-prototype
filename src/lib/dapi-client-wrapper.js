import DAPIClient from '@dashevo/dapi-client'

export class DapiClientWrapper {
  constructor(seeds) {
    this.dapiClient = new DAPIClient({
      seeds,
      timeout: 30000
    })
  }

  signUpUser = async privateKey => {
    const userPrivateKey = privateKey ? privateKey : new PrivateKey()
    const validPayload = new Transaction.Payload.SubTxRegisterPayload()
      .setUserName(username)
      .setPubKeyIdFromPrivateKey(userPrivateKey)
      .sign(userPrivateKey)

    const { items: inputs } = await this.dapiClient.getUTXO(this.faucetAddress)

    const transaction = Transaction()
      .setType(Transaction.TYPES.TRANSACTION_SUBTX_REGISTER)
      .setExtraPayload(validPayload)
      .from(inputs.slice(-1)[0])
      .addFundingOutput(10000)
      .change(this.faucetAddress)
      .sign(this.faucetPrivateKey)

    return await this.dapiClient.sendRawTransaction(transaction.serialize())
  }

  publishUserProfile = async stPacket => {
    // 1. Create ST Packet

    // throw new Error(JSON.stringify(stPacket, null, 2))

    // 2. Create State Transition
    const transaction = new Transaction().setType(Transaction.TYPES.TRANSACTION_SUBTX_TRANSITION)

    transaction.extraPayload
      .setRegTxId(regTxId)
      .setHashPrevSubTx(regTxId)
      .setHashSTPacket(stPacket.hash())
      .setCreditFee(1000)
      .sign(userPrivateKey)

    const transitionHash = await this.dapiClient.sendRawTransition(
      transaction.serialize(),
      stPacket.serialize().toString('hex')
    )

    if (!transitionHash) {
      throw new Error('Could not publish user profile')
    }
  }
}
