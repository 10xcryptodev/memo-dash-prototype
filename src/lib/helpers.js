export const filterUser = (username, users) => {
  if (!users || users.length === 0) return undefined
  return users.find(user => user.username === username)
}

export const isMemoLikedByUsername = (memo, username, users) => {
  const currentUser = filterUser(username, users)
  if (!currentUser) return false

  return currentUser.ownLikes.some(like => memo.username !== username && like.relation.index === memo.idx)
}
