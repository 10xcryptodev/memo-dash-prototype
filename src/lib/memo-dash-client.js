import _ from 'lodash'

function proxy(type, def) {
  // Bypass proxy for now...
  return new Proxy({}, {
    get(target, name) {
      // Promises will always check to see if this prop exists
      if (name === 'then') {
        return undefined
      }

      if (_.has(def, name)) {
        return def[name]
      } else {
        console.error('NEED PROP', type + '.' + name)
      }

      return 'TEMPVAL'
    },
    set(target, name, value) {
      if (def.name === 'bob' && name === 'username') {
        debugger
      }

      def[name] = value
      return true
    }
  })
}

function NewUser(props) {
  // return proxy('user', new User(props))
  return new User(props)
}

function NewDAP(props) {
  // return proxy('dap', new DAP(props))
  return new DAP(props)
}

function NewSignup(props) {
  // return proxy('signup', new Signup(props))
  return new Signup(props)
}

function NewMemo(props) {
  // return proxy('memo', new Memo(props))
  return new Memo(props)
}

class DAP {
  constructor(schema) {
    _.extend(this, schema)
    this.id = this.$id
  }
}

class User {
  constructor(username) {
    this.id = _.uniqueId('user')
    this.username = username
    this.name = username
    this.signups = []
    this.following = []
    this.followers = []
    this.likes = []
    this.profile = proxy('profile', {
      username: username,
      bio: 'This mysterious user prefers to remain an enigma.',
      address: 'Their location is undisclosed',
      avatarUrl: undefined,
      followersCount: 0,
      followingCount: 0,
      likesCount: 0,
    })
  }
}

class Signup {
  constructor(data) {
    _.extend(this, data)
  }
}

class Memo {
  constructor(data) {
    _.extend(this, data)
    this.replies = []
  }
}

function filterDeleted(memo) {
  return !memo.deleted
}

function filterReplies(memo) {
  return !memo.parentTxi
}

export default class MemoDashClient {
  daps = []
  users = []
  memos = []
  currentUser = undefined
  currentDap = undefined


  constructor() {
    console.log('stub')
  }

  createBlockchainUser(username) {
    console.log('@createBlockchainUser', Array.from(arguments).slice(-2))
    // console.log('@createBlockchainUser', username)
    // console.log('create user', username)
    const user = NewUser(username)
    this.users.push(user)
    return Promise.resolve(username)
  }

  login(username) {
    console.log('@login', Array.from(arguments).slice(-2))
    if (typeof username !== 'string') {
      username = username.blockchainUsername
    }

    let user = _.find(this.users, {username})
    if (user === undefined) {
      console.error('user not found', username, this.users)
      return Promise.reject('user not found')
    }
    this.currentUser = user
    return Promise.resolve(user)
  }

  logout() {
    console.log('@logout', Array.from(arguments).slice(-2))
    if (this.currentUser === undefined) {
      console.error('Cannot logout - no current user')
      return Promise.reject()
    }
    delete this.currentUser
    return Promise.resolve()
  }

  searchDaps(title) {
    console.log('@searchDaps', Array.from(arguments).slice(-2))
    return Promise.resolve(_.filter(this.daps, {title}))
  }

  createDap(schema) {
    console.log('@createDap', Array.from(arguments).slice(-2))
    const dap = NewDAP(schema)
    this.daps.push(dap)
    return Promise.resolve(dap.id)
  }

  getDap(dapid) {
    console.log('@getDap', Array.from(arguments).slice(-2))
    return Promise.resolve(_.find(this.daps, {id: dapid}))
  }

  setDap(dap) {
    console.log('@setDap', Array.from(arguments).slice(-2))
    this.currentDap = dap
  }

  signup({text, avatarUrl, address, name}) {
    console.log('@signup', Array.from(arguments).slice(-2))
    // const user = _.find(this.users, {a: this.currentUser})
    // console.log('new signup', text, avatarUrl, address, name)
    // this.currentUser.signups.push({
    //   text,
    //   avatarUrl,
    //   address,
    //   name,
    // })

    this.currentUser.profile.bio = text
    this.currentUser.profile.avatarUrl = avatarUrl
    this.currentUser.profile.address = address
    this.currentUser.profile.name = name

    return Promise.resolve()
  }

  log() {
    console.log.apply(console, arguments)
  }

  searchUsers(username) {
    console.log('@searchusers', username)
    const user = _.find(this.users, {username})
    return Promise.resolve([user])
  }

  getAllProfiles() {
    // No args - current user?
    console.log('@getAllProfiles', Array.from(arguments).slice(-2))
    const allProfiles = this.users.map(u => u.profile)
    return Promise.resolve(allProfiles)
  }

  getUserProfile(username) {
    console.log('@getUserProfile', username)
    // const profile = proxy('userprofile', {})
    const user = _.find(this.users, {username})
    if (user === undefined) {
      debugger
      console.error('user not found')
      return Promise.reject()
    }
    console.log(user.profile.bio)
    return Promise.resolve(user.profile)
  }

  updateProfile({text}) {
    console.log('@updateProfile', Array.from(arguments).slice(-2))
    this.currentUser.profile.bio = text
    return Promise.resolve()
  }

  getUserId(username) {
    console.log('@getUserId', Array.from(arguments).slice(-2))
    const targetUser = _.find(this.users, {username})
    if (targetUser === undefined) {
      console.error('user not found')
      return Promise.reject('user not found')
    }
    return Promise.resolve(targetUser.id)
  }

  followUser(username) {
    console.log('@followUser', Array.from(arguments).slice(-2))
    // console.log(this.currentUser.username, 'is following user', username)

    const targetUser = _.find(this.users, {username})
    if (targetUser === undefined) {
      console.error('user not found')
      return Promise.reject('user not found')
    }

    this.currentUser.following.push(targetUser)
    this.currentUser.profile.followingCount++
    targetUser.followers.push(this.currentUser)
    targetUser.profile.followersCount++

    return Promise.resolve()
  }

  unFollowUser(username) {
    const targetUser = _.find(this.users, {username})
    if (targetUser === undefined) {
      console.error('user not found')
      return Promise.reject('user not found')
    }

    this.currentUser.following = _.without(this.currentUser.following, targetUser)
    this.currentUser.profile.followingCount--
    targetUser.followers = _.without(targetUser.followers, this.currentUser)
    targetUser.profile.followersCount--

    return Promise.resolve()
  }

  getUserFollowers(username) {
    console.log('@getUserFollowers', Array.from(arguments).slice(-2))
    const user = _.find(this.users, {username})
    if (user === undefined) {
      console.error('user not found')
      return Promise.reject()
    }
    return Promise.resolve(user.followers)
  }

  getUserFollowing(username) {
    console.log('@getUserFollowing', Array.from(arguments).slice(-2))
    const user = _.find(this.users, {username})
    if (user === undefined) {
      console.error('user not found')
      return Promise.reject()
    }
    return Promise.resolve(user.following)
  }

  getUserLikes(username) {
    console.log('@getUserLikes', Array.from(arguments).slice(-2))

    const user = _.find(this.users, {username})
    if (user === undefined) {
      console.error('User not found')
      return Promise.reject('User not found')
    }

    return Promise.resolve(user.likes)
  }

  getUsername(userId) {
    console.log('@getUsername', Array.from(arguments).slice(-2))
    const user = _.find(this.users, {id: userId})
    return Promise.resolve(user.username)
  }

  getMemo(username, memoId) {
    console.log('@getMemo', Array.from(arguments).slice(-2))

    const memo = _.find(this.memos, {username, idx: memoId})
    if (memo === undefined) {
      console.error('Memo not found')
      return Promise.reject('Memo not found')
    }

    return Promise.resolve(memo)
  }

  getMemosByUsername(username) {
    const memos = this.memos.filter(m => m.username === username).filter(filterDeleted).filter(filterReplies)
    console.log('@getMemosByUsername', Array.from(arguments).slice(-2), memos)
    return Promise.resolve(memos)
  }

  getMemos() {
    const memos = this.memos.filter(filterDeleted).filter(filterReplies)
    console.log('@getMemos', Array.from(arguments).slice(-2), memos)
    return Promise.resolve(memos)
  }

  postMemo(memoText) {
    console.log('@postMemo', Array.from(arguments).slice(-2))

    const memo = NewMemo({
      username: this.currentUser.username,
      idx: _.uniqueId('memo'),
      memoRepliesCount: 0,
      memoLikesCount: 0,
      memoTipTotal: 0,
      memoDatetime: new Date(),
      memoText,
    })

    this.memos.push(memo)
    return Promise.resolve()
  }

  deleteMemo(memoId) {
    console.log('@deleteMemo', Array.from(arguments).slice(-2))

    const memo = _.find(this.memos, {idx:memoId})
    // this.memos = _.without(this.memos, memo)
    memo.deleted = true

    if (memo.parentIdx) {
      const parentMemo = _.find(this.memos, {idx: memo.parentIdx})
      parentMemo.memoRepliesCount--
    }

    return Promise.resolve()
  }

  likeMemo(username, memoId) {
    console.log('@likeMemo', Array.from(arguments).slice(-2))

    const memo = _.find(this.memos, {username, idx: memoId})
    if (memo === undefined) {
      console.error('memo not found')
      return Promise.reject('memo not found')
    }

    const user = _.find(this.users, {username})
    if (user === undefined) {
      console.error('user not found')
      return Promise.reject('user not found')
    }

    user.profile.likesCount++
    memo.memoLikesCount++

    this.currentUser.likes.push(proxy('like', {
      username: this.currentUser.username,
      userId: this.currentUser.username,
      idx: _.uniqueId('like'),
      relation: proxy('relation', {
        userId: user.id,
        index: memo.idx
      }),
    }))

    return Promise.resolve(memo)
  }

  removeLike(likeId) {
    console.log('@removeLike', Array.from(arguments).slice(-2))

    const like = _.find(this.currentUser.likes, {idx: likeId})
    if (like === undefined) {
      console.error('like not found')
      return Promise.reject('like not found')
    }

    const user = _.find(this.users, {id: like.relation.userId})
    if (user === undefined) {
      console.error('user not found')
      return Promise.reject('user not found')
    }

    const memo = _.find(this.memos, {idx: like.relation.index})
    if (memo === undefined) {
      console.error('memo not found')
      return Promise.reject('memo not found')
    }

    user.profile.likesCount--
    memo.memoLikesCount--

    this.currentUser.likes = _.without(this.currentUser.likes, like)

    return Promise.resolve()
  }

  editMemo(memoId, text) {
    console.log('@editMemo', Array.from(arguments).slice(-2))

    const memo = _.find(this.memos, {idx: memoId})
    if (memo === undefined) {
      console.error('Memo not found')
      return Promise.reject('Memo not found')
    }

    memo.memoText = text

    return Promise.resolve()
  }

  replyToMemo(username, memoId, memoText) {
    console.log('@replyToMemo', Array.from(arguments).slice(-2))

    const user = _.find(this.users, {username})
    if (user === undefined) {
      console.error('User not found')
      return Promise.reject('User not found')
    }

    const memo = _.find(this.memos, {idx: memoId})
    if (memo === undefined) {
      console.error('Memo not found')
      return Promise.reject('Memo not found')
    }

    const reply = NewMemo({
      username: this.currentUser.username,
      idx: _.uniqueId('memo'),
      memoRepliesCount: 0,
      memoLikesCount: 0,
      memoTipTotal: 0,
      memoDatetime: new Date(),
      memoText,
      parentIdx: memo.idx // made up by me, do they use something like this?
    })

    this.memos.push(reply)
    memo.memoRepliesCount++
    memo.replies.push({username, idx: reply.idx})

    return Promise.resolve()
  }

  getMemoReplies(username, memoId) {
    console.log('@getMemoReplies', Array.from(arguments).slice(-2))

    const memo = _.find(this.memos, {idx: memoId})
    if (memo === undefined) {
      console.error('memo not found')
      return Promise.reject('memo not found')
    }
    const replies = memo.replies.map(ref => _.find(this.memos, {idx: ref.idx})).filter(filterDeleted)
    console.log('REPLIES', replies)

    return Promise.resolve(replies)
  }
}