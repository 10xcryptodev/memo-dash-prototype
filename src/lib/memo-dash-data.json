{
  "alice_subtx_1": {
    "subtx": {
      "pver": 1,
      "action": 1,
      "uname": "alice",
      "pubkey": "024964f06ea5cfec1890d7e526653b083c12360f79164c1e8163327d0849fa7bca",
      "meta": {
        "id": "2d30902738f48f6f927a296bfbf058f1e09ebb201ca3c1a6c462f2bb2ee05705",
        "sig": "INxlyf9JX8j1hfcBNf5W72+UALu7+5nF8l/MfUZSJlomNDkmwWfvva4eE4/tjJPUO+ByV6K3cbUgPhjbEZIM8Ik="
      }
    }
  },
  "bob_subtx_1": {
    "subtx": {
      "pver": 1,
      "action": 1,
      "uname": "bob",
      "pubkey": "0261f767fb9c0bad4ea5425dbb618522ae4ea186beb37fb1b0666c59a5acd05b93",
      "meta": {
        "id": "76b19e926e6ea960ce19a9ba7db205b3c5f05c178240348e4236abb821e9c261",
        "sig": "H3bcSB3TkT/GOC6dFsZ1eS3R3k0RR53DEBcytscUovXbBW2Crys+q4Y1EgbqtttduEnYs8sUcymhLZXh/bk5DBs="
      }
    }
  }
}