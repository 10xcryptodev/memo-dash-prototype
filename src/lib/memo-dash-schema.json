{
  "$id": "http://dash.org/schema/dap",
  "title": "DashMemo",
  "profile": {
    "type": "object",
    "allOf": [
      {
        "$ref": "http://dash.org/schemas/sys#/definitions/dapobjectbase"
      }
    ],
    "primaryKey": {
      "composite": true
    },
    "properties": {
      "name": {
        "type": "string",
        "minLength": 1,
        "maxLength": 144
      },
      "address": {
        "type": "string"
      },
      "text": {
        "type": "string",
        "minLength": 1,
        "maxLength": 144
      },
      "avatarUrl": {
        "type": "string",
        "format": "uri"
      }
    },
    "required": [
      "address"
    ]
  },
  "memo": {
    "type": "object",
    "allOf": [
      {
        "$ref": "http://dash.org/schemas/sys#/definitions/dapobjectbase"
      }
    ],
    "properties": {
      "message": {
        "type": "string",
        "minLength": 1,
        "maxLength": 144
      },
      "createdAt": {
        "type": "string",
        "format": "date-time"
      },
      "updateAt": {
        "type": "string",
        "format": "date-time"
      },
      "relation": {
        "$ref": "http://dash.org/schemas/sys#/definitions/relation"
      }
    },
    "required": [
      "message",
      "createdAt"
    ]
  },
  "like": {
    "type": "object",
    "allOf": [
      {
        "$ref": "http://dash.org/schemas/sys#/definitions/dapobjectbase"
      }
    ],
    "primaryKey": {
      "composite": true,
      "includes": ["relation"]
    },
    "properties": {
      "relation": {
        "$ref": "http://dash.org/schemas/sys#/definitions/relation"
      },
      "tipTxHash": {
        "type": "string"
      }
    },
    "required": [
      "relation"
    ]
  },
  "follow": {
    "type": "object",
    "allOf": [
      {
        "$ref": "http://dash.org/schemas/sys#/definitions/dapobjectbase"
      }
    ],
    "primaryKey": {
      "composite": true,
      "includes": ["relation"]
    },
    "properties": {
      "relation": {
        "$ref": "http://dash.org/schemas/sys#/definitions/relation",
        "isPrimary": true
      }
    },
    "required": [
      "relation"
    ]
  }
}