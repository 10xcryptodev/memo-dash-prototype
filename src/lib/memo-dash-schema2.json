{
  "$id": "https://schema.dash.org/dpp-0-4-0/memo/memo",
  "properties": {
    "id": {
      "type": "string",
      "minLength": 42,
      "maxLength": 44,
      "pattern": "^[123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]+$"
    },
    "title": {
      "type": "string",
      "minLength": 1,
      "maxLength": 50
    }
  },
  "required": [
    "id",
    "title"
  ]
}
