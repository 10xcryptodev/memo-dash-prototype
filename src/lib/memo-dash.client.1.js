import DAPIClient from "@dashevo/dapi-client";
import DashPlatformProtocol from "@dashevo/dpp";
import { PrivateKey, PublicKey, Address } from "@dashevo/dashcore-lib";

import memoDashContract from "./memo-dash.contract";

export class MemoDashClient {
  constructor(networkType, seeds, privateKey) {
    this.dpp = new DashPlatformProtocol();
    this.dapiClient = new DAPIClient({
      seeds,
      timeout: 30000
    });

    faucetPrivateKey = new PrivateKey(privateKey);
    const faucetPublicKey = PublicKey.fromPrivateKey(faucetPrivateKey);
    faucetAddress = Address.fromPublicKey(
      faucetPublicKey,
      networkType === "devnet" ? "testnet" : networkType
    ).toString();

    const contract = dpp.contract.create("memo-dash", memoDashContract);

    const result = dpp.contract.validate(contract);
    console.log(`MemoDash contract is valid: ${result.isValid()}`);
    dpp.setContract(contract);
  }

  /**
   * Returns user id for the given username.
   * @param {string} username
   * @return {Promise<string>}
   */
  async getUserId(username) {
    const user = await this.dapiClient.getUserByName(username);
    return user ? user.uname : undefined;
  }

  /**
   * Return username by id
   * @param {string} userId
   * @return {Promise<string>}
   */
  async getUsername(userId) {
    const user = await this.dapiClient.getUserById(userId);
    return user ? user.uname : undefined;
  }

  /**
   * Get all Memos
   * @returns
   * [{
   *   username,
   *   memoDatetime,
   *   memoText,
   *   memoLikesCount,
   *   memoTipTotal,
   *   memoRepliesCount
   * }]
   */
  async getMemos() {}

  /**
   * @return {Promise<[]>}
   */
  async getAllProfiles() {}

  /**
   *  ------------------------
   *  Queries
   *  ------------------------
   */

  /**
   * Get a user's profile info
   *
   * returns info for profile page on a MemoDash user
   * e.g. https://memo.cash/profile/19RyV6XQEww5td2LPWDpK8o5V8at7Vpwgv
   * @param username
   * @returns
   * {
   *   username,
   *   bio,
   *   avatarUrl,
   *   followersCount,
   *   followingCount,
   *   likesCount
   * }
   */
  async getUserProfile(username) {}

  async getOwnProfile() {}

  /**
   * Get a specific Memo
   * @param {string} username
   * @param {number} idx
   * @returns
   * Promise<{
   *   username,
   *   memoDatetime,
   *   memoText,
   *   memoLikesCount,
   *   memoTipTotal,
   *   memoRepliesCount
   * }>
   */
  async getMemo(username, idx) {}

  /**
   * @param {string} username
   * @return {Promise<Array<{
   *   username,
   *   memoDatetime,
   *   memoText,
   *   memoLikesCount,
   *   memoTipTotal,
   *   memoRepliesCount
   * }>>}
   */
  async getMemosByUsername(username) {}

  /**
   * Returns all own memos
   * @return {Promise<Array<{username, memoDatetime, memoText, memoLikesCount, memoTipTotal, memoRepliesCount}>>}
   */
  async getAllOwnMemos() {}

  /**
   * Edit own memo at the given infex
   * @param {string} idx
   * @param {string} newMessage
   * @return {Promise<boolean>}
   */
  async editMemo(idx, newMessage) {}

  /**
   * Removes memo context
   * @param {string} idx
   * @return {Promise<boolean>}
   */
  async deleteMemo(idx) {}

  /**
   * Get Memos that reply to a given Memo
   * @returns
   * [{
   *   username,
   *   memoDatetime,
   *   memoText,
   *   memoLikesCount,
   *   memoTipTotal,
   *   memoRepliesCount
   * }]
   */
  async getMemoReplies(username, idx) {}

  /**
   *  ------------------------
   *  Actions
   *  ------------------------
   */

  /**
   * Create memo dash profile
   * @param profileData
   * @param {string} profileData.text
   * @param {string} profileData.avatarUrl
   * @param {string} profileData.address
   * @param {string} profileData.name
   * @return {Promise<*>}
   */
  async signup(profileData) {}

  /**
   * Update memo dash profile
   * @param newProfileData
   * @param {string} [newProfileData.text]
   * @param {string} [newProfileData.avatarUrl]
   * @param {string} [newProfileData.address]
   * @param {string} [newProfileData.name]
   * @return {Promise<boolean>}
   */
  async updateProfile(newProfileData) {}

  /**
   * Post memo to your profile.
   * @param {string} message
   * @return {Promise<boolean>}
   */
  async postMemo(message) {}

  /**
   * Follow a user
   * @param {string} username
   * @return {Promise<boolean>}
   */
  async followUser(username) {}

  /**
   * Get followers of a user
   * @returns
   * [{
   *   username
   * }]
   */
  async getUserFollowers(username) {
    const { related } = await this.getDapContext(username);
    if (related) {
      return this.filterObjectsByType(related, "follow").map(follower => ({
        username: follower.meta.uname
      }));
    }
    return [];
  }

  /**
   * Get users followed by a user
   * @returns
   * [{
   *   username
   * }]
   */
  async getUserFollowing(username) {}

  /**
   * Unfollow user
   * @param {string} username
   * @return {Promise<boolean>}
   */
  async unFollowUser(username) {}

  async replyToMemo(username, idx, message) {}

  /**
   * Like memo
   * @param {string} username - username of user who posted the memo
   * @param {string} idx - memo id
   * @return {Promise<*>}
   */
  async likeMemo(username, idx) {}

  async getAllOwnLikes() {}

  /**
   * Remove like from memo
   * @param {string} idx - like id
   * @return {Promise<void>}
   */
  async removeLike(idx) {}

  async tipMemo(username, idx, amount) {}

  /**
   *  ------------------------
   *  Not required (from memo.cash)
   *  ------------------------
   *
   *  Set image base URL
   *  Attach picture
   *  Set profile picture
   *  Repost memo
   *  Post topic message
   *
   */
}
