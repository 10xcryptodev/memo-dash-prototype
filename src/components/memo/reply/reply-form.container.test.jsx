import React from 'react'
import { shallow } from 'enzyme'
import ReplyFormContainer from './reply-form.container'
import { filterUser } from '../../../lib/helpers'

jest.mock('../../../lib/helpers')

describe('<ReplyFormContainer />', () => {
  let testUser
  beforeEach(() => {
    testUser = filterUser('alice')
  })

  describe('Shallow rendering', () => {
    it('renders without crashing', () => {
      const wrapper = shallow(
        <ReplyFormContainer username={testUser.username} memoId={testUser.memos[0].idx} />
      )
      expect(wrapper).toMatchSnapshot()
    })
  })
})
