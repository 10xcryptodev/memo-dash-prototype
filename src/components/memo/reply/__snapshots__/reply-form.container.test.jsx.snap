// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReplyFormContainer /> Shallow rendering renders without crashing 1`] = `
<Formik
  enableReinitialize={false}
  handleSubmit={[Function]}
  initialValues={
    Object {
      "memoId": 1,
      "message": "",
      "username": "alice",
    }
  }
  isInitialValid={false}
  memoId={1}
  onSubmit={[Function]}
  render={[Function]}
  username="alice"
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;
